pkg/packageinstall/packageinstall.go

```diff
@@ -266,3 +278,62 @@ func (pi *PackageInstallCR) update(updateFunc func(*pkgingv1alpha1.PackageInstal
 
 	return fmt.Errorf("Updating package install: %s", lastErr)
 }
+
+func (pi *PackageInstallCR) createPlaceHolderSecrets(pv *datapkgingv1alpha1.Package) error {
+	for i, fetch := range pv.Spec.Template.Spec.Fetch {
+		if fetch.ImgpkgBundle != nil && fetch.ImgpkgBundle.SecretRef == nil {
+			secretName, err := pi.createSecretForSecretgenController(i)
+			if err != nil {
+				return err
+			}
+			pv.Spec.Template.Spec.Fetch[i].ImgpkgBundle.SecretRef = &kcv1alpha1.AppFetchLocalRef{secretName}
+		}
+
+		if fetch.Image != nil && fetch.Image.SecretRef == nil {
+			secretName, err := pi.createSecretForSecretgenController(i)
+			if err != nil {
+				return err
+			}
+			pv.Spec.Template.Spec.Fetch[i].Image.SecretRef = &kcv1alpha1.AppFetchLocalRef{secretName}
+		}
+	}
+	return nil
+}
+
+func (pi PackageInstallCR) createSecretForSecretgenController(iteration int) (string, error) {
+	secretName := fmt.Sprintf("%s-%s", pi.model.Name, "fetch"+strconv.Itoa(iteration))
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      secretName,
+			Namespace: pi.model.Namespace,
+			Annotations: map[string]string{
+				"secretgen.carvel.dev/image-pull-secret": "",
+			},
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: "packaging.carvel.dev/v1alpha1",
+					Kind:       "PackageInstall",
+					Name:       pi.model.Name,
+					UID:        pi.model.UID,
+				},
+			},
+		},
+		Data: map[string][]byte{
+			corev1.DockerConfigJsonKey: []byte(`{"auths":{}}`),
+		},
+		Type: corev1.SecretTypeDockerConfigJson,
+	}
+
+	_, err := pi.coreClient.CoreV1().Secrets(pi.model.Namespace).Create(context.Background(), secret, metav1.CreateOptions{})
+	if err != nil {
+		if errors.IsAlreadyExists(err) {
+			_, err = pi.coreClient.CoreV1().Secrets(pi.model.Namespace).Update(context.Background(), secret, metav1.UpdateOptions{})
```

Comment:
should we bother with updates right now? one challenge is that secret updates are sensitive to type changes -- not allowed, so it would have to be delete/create cycle.

Reply:
The issue I forsee with not including is how we deal with updates to secrets after the PackageInstall is initially created? What is our recommended flow for having the PackageInstall's placeholder secrets include secret updates?